### PR TITLE
Bug-fixes and upgrades to the Active Space Module

### PR DESCRIPTION
**Pull request by mrossinek**
# Summary
In this PR, we are upgrading the active space module of the Quickstep method.
Here is a short overview of the upgrades:
- bug fixes of some incorrect indices
- extension to support UKS calculations
- implementation of an FCIDump checksum to ease testing
- addition of regression tests
- minor documentation improvements

## Details
After this PR, the active space module can be used in conjunction with HF-style calculations. This includes:
- GAPW with the ALL electron potential
- GAPW with the GTH-HF potential
- GPW with the GTH-HF potential

Regression tests for all of the above scenarios have been added in `tests/QS/regtest-as-1`.
```
% ../tools/regtesting/do_regtest -cp2kdir ../ -version sdbg -arch local -nobuild -restrictdir QS/regtest-as-1
<...>
--------------------------------- Summary --------------------------------
Number of FAILED  tests 0
Number of WRONG   tests 0
Number of CORRECT tests 26
Total number of   tests 26
GREPME 0 0 26 0 26 X

Summary: correct: 26 / 26; 3min
Status: OK
```

## Credits
Authored by Valery Weber and Max Rossmannek.

### CODE REVIEW

**Comment by mrossinek** [comment]
I rebased the branch off of `master`, in order to remove the merge commits. I will force push this branch in a second so in case anyone already pulled from it, please be aware of that.

File: src/qs_active_space_methods.F [hunk 1]
```diff
@@ -1856,10 +2140,41 @@ LOGICAL FUNCTION eri_fcidump_print_func(this, i, j, k, l, val) RESULT(cont)
 
       ! write to the actual file only on the master
       IF (this%unit_nr > 0) THEN
-         WRITE (this%unit_nr, "(ES23.16,4I4)") val, i, j, k, l
+         !WRITE (this%unit_nr, "(ES23.16,4I4)") val, i, j, &
+         !     &                                     k, l
```
Comment (dev-zero) [inline src/qs_active_space_methods.F]: please remove commented out code completely, unless required later

File: src/qs_active_space_methods.F [hunk 2]
```diff
@@ -1856,10 +2140,41 @@ LOGICAL FUNCTION eri_fcidump_print_func(this, i, j, k, l, val) RESULT(cont)
...
+      INTEGER, INTENT(in)                     :: i, j, k, l
+      REAL(KIND=dp), INTENT(in)               :: val
+      ! TODO Without performing some no-operation on these arguments, the compiler complains about
+      ! unused dummy arguments. However, we cannot remove them in the current implementation. How
+      ! should we handle this clearnly?
+#define nop(x) associate( x => x ); END associate
+      nop(i)
+      nop(j)
+      nop(k)
+      nop(l)
```
Comment (dev-zero) [inline src/qs_active_space_methods.F]: there is already a macro available: `MARK_USED(i)`

**Comment by dev-zero** [comment]
@mrossinek: besides the 2 remarks, looks good to me. Commits will be squashed into one unless you want to do an interactive rebase and squash them into a small number of commits doing isolated changes each.

**Comment by mrossinek** [comment]
Hi @dev-zero, Thanks for the input! I implemented both of your suggested changes.

I would favor a squashed merge for the sake of time. You can use the description of this PR for the commit message as it probably provides a more comprehensive overview of the contents than a concatenation of the squashed commit messages.

EDIT: do I need to rebase this branch off of master again or can we merge it without doing that again? As long as there are no conflicts it should be fine like this, right?

**Comment by dev-zero** [comment]
@mrossinek thanks, will do! No, unless there's a message saying that it can't be merged/rebased automatically you don't have to rebase.

**Comment by mrossinek** [comment]
I received an email from the CP2K dashboard saying that the AiiDA Plugin broke because of this PR: https://dashboard.cp2k.org/archive/aiida/index.html

I have not heard of AiiDA before so any pointers as to what I need to do about this would be highly appreciated :slightly_smiling_face:

**Comment by dev-zero** [comment]
the AiiDA failure has likely more to do with the recent changes to aiida-cp2k itself rather than this PR

**Comment by dev-zero** [comment]
> the AiiDA failure has likely more to do with the recent changes to aiida-cp2k itself rather than this PR

most likely [this one](https://github.com/aiidateam/aiida-cp2k/commit/090ca4987871632e70c5b0fab619d8b3902a1e18) which introduced the failing code part

**Comment by mrossinek** [comment]
Okay good to know.
The `ssmp` toolchain failed too, though. But it was in an unrelated regtest:
```
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
/workspace/cp2k/regtesting/local/ssmp/TEST-local-ssmp-2020-12-18_12-38-33/QMMM/QS/regtest-1/H2O-qmmm-gauss-6.inp.out : 
 Total energy: : ref = -16.78375982393304 new = -16.78375982393656  
 relative error :   2.09770593e-13 >  numerical tolerance = 2e-13  
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx

<...>

--------------------------------- Summary --------------------------------
Number of FAILED  tests 0
Number of WRONG   tests 1
Number of CORRECT tests 3366
Total number of   tests 3367
GREPME 0 1 3366 0 3367 X

Summary: correct: 3366 / 3367; wrong: 1; 16min
Status: FAILED

--------------------------------------------------------------------------
Regtest took 958.00 seconds.
--------------------------------------------------------------------------
Fri Dec 18 12:54:30 UTC 2020
*************************** testing ended ********************************
```

**Pull request merged**